lib/maths/common/CTools.cc

```diff
@@ -1393,63 +1399,88 @@ operator()(const lognormal& logNormal, double a, double b) const {
     if (a > b) {
         std::swap(a, b);
     }
-    if (a == POS_INF) {
+    if (a >= POS_INF) {
         return POS_INF;
     }
     if (b <= 0.0) {
         return 0.0;
     }
+    if (a == b) {
+        return a;
+    }
 
     double location = logNormal.location();
     double scale = logNormal.scale();
+    if (a <= 1e-8) {
+        lognormal logNormalShift{location + pow2(scale), scale};
+        // If F(b) underflows we're in the extreme left tail. We can fallback to
+        // computing an asymptotic expansion of the ratio. The leading term (by
+        // L'Hospital) turns out to just be b. This intuitively makes sense since
+        // the distribution is concentrated at the right end of the interval.
+        double cdf = boost::math::cdf(logNormal, b);
+        return (cdf == 0.0 ? b
+                           : boost::math::mean(logNormal) *
+                                 boost::math::cdf(logNormalShift, b) / cdf);
+    }
+
     double mean = boost::math::mean(logNormal);
-    double loga = a <= 0.0 ? NEG_INF : std::log(a);
-    double logb = b == POS_INF ? POS_INF : std::log(b);
-    double c = location + scale * scale;
+    double loga = std::log(a);
+    double logb = std::log(b);
+    double c = location + pow2(scale);
     double s = std::sqrt(2.0) * scale;
-    double a_ = loga == NEG_INF ? NEG_INF : (loga - location) / s;
-    double b_ = logb == POS_INF ? POS_INF : (logb - location) / s;
-    double erfa = loga == NEG_INF ? -1.0 : std::erf((loga - c) / s);
-    double erfb = logb == POS_INF ? 1.0 : std::erf((logb - c) / s);
+    double a_ = (loga - location) / s;
+    double b_ = (logb - location) / s;
+    double erfa = std::erf((loga - c) / s);
+    double erfb = std::erf((logb - c) / s);
 
     if (erfb - erfa < std::sqrt(EPSILON)) {
-        double expa = loga == NEG_INF ? 0.0 : std::exp(-a_ * a_);
-        double expb = logb == POS_INF ? 0.0 : std::exp(-b_ * b_);
+        double expa = std::exp(-a_ * a_);
+        double expb = std::exp(-b_ * b_);
         return expa == expb ? (2.0 * a / (a + b)) * b
                             : (expa + expb) / (expa / a + expb / b);
     }
 
-    double erfa_ = a_ == NEG_INF ? -1.0 : std::erf(a_);
-    double erfb_ = b_ == POS_INF ? 1.0 : std::erf(b_);
-    return mean * (erfb - erfa) / (erfb_ - erfa_);
+    double erfa_ = std::erf(a_);
+    double erfb_ = std::erf(b_);
+    return (erfb - erfa) == (erfb_ - erfa_) ? mean : mean * (erfb - erfa) / (erfb_ - erfa_);
```

Comment:
Is it possible that `erfb_ - erfa_` could be zero when `erfb - erfa` is not zero?

Reply:
In this case the answer should be inf (at least to working precision). But maybe though I should actually trap to max double because inf probably creates downstream issues. It shouldn't really happen for realistic data but better safe than sorry.

Reply:
See [cbaa29e](https://github.com/elastic/ml-cpp/pull/2286/commits/cbaa29e7862fec6da97f6fb5233edb3582b5c4ed).